parquet/src/main/java/org/apache/iceberg/parquet/ParquetDictionaryRowGroupFilter.java

```diff
@@ -108,11 +118,15 @@ private boolean eval(MessageType fileSchema, BlockMetaData rowGroup,
       }
 
       for (ColumnChunkMetaData meta : rowGroup.getColumns()) {
-        PrimitiveType colType = fileSchema.getType(meta.getPath().toArray()).asPrimitiveType();
-        if (colType.getId() != null) {
-          int id = colType.getId().intValue();
-          isFallback.put(id, ParquetUtil.hasNonDictionaryPages(meta));
-          mayContainNulls.put(id, mayContainNull(meta));
+        try {
+          PrimitiveType colType = fileSchema.getType(meta.getPath().toArray()).asPrimitiveType();
+          if (colType.getId() != null) {
+            int id = colType.getId().intValue();
+            isFallback.put(id, ParquetUtil.hasNonDictionaryPages(meta));
+            mayContainNulls.put(id, mayContainNull(meta));
+          }
+        } catch (org.apache.parquet.io.InvalidRecordException e) {
```

Comment:
Do you have a test case where this is thrown?

Reply:
Yes. The `testImportWithIncompatibleSchema` in TestSparkTableUtil.java.

Reply:
The problem here is that `ApplyNameMapping` is omitting fields that are in the file from the schema with field IDs. Instead of changing these filters, `ApplyNameMapping` should return all fields of the file schema, even if they are not assigned field IDs.

By doing that, we can assume that the file schema always matches the file metadata. That's important to avoid unexpected exceptions like this one: we're using the file's column metadata to look up a column in the schema and that should always work.

Reply:
@chenjunjiedada, this needs to be fixed so that there are no changes to the filter implementations.

Reply:
I tried to remove the `try...catch...` here and let `ApplyNameMapping` to return the schema with partial IDs assigned. That causes problems when we visit the parquet type.

Firstly, the `PruneColumns` visitor assumes the `typeWithIds` is fully assigned with IDs. Its `getId` has a precondition check. A file schema with partially assigned IDs cannot pass the condition check.

Secondly, the `HasId` logic also has the same assumption, which makes `SparkParquetReader` throws NPE when it calls `filedType.getId().intValue()`.

Reply:
`HasId` should definitely succeed if there are fields without them. Its purpose is to check and we expect it to encounter schemas without IDs. Why is `HasId` being run on a schema that has had a name mapping applied?

If `PruneColumns` assumes all IDs are there, then I think we should fix that assumption.

Reply:
`SparkParquetReader` build parquet reader according to `typeWithIds` passed from ReadConf.java. The `HasId` is used in ReadConf.java to assign IDs to file schema, and then it is used in `SparkParquetReader` to determine whether passing files schema has assigned IDs.  

Besides `PruneColumsn`, The `ReadBuilder` also assumes all IDs assigned.

Reply:
@rdblue , do you think we need to change both `PruneColumns` and `SparkParquetReader#ReadBuilder` to satisfy the assumption? It may change a lot.